sycl/include/CL/sycl/marray.hpp

```diff
@@ -0,0 +1,355 @@
+//==----------------- marray.hpp --- SYCL types ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// Implements marray classes.
+
+#pragma once
+
+#include <CL/sycl/aliases.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/generic_type_traits.hpp>
+#include <CL/sycl/detail/type_traits.hpp>
+#include <CL/sycl/half_type.hpp>
+
+#include <array>
+#include <cmath>
+#include <cstring>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+/// Provides a cross-patform math array class template that works on
+/// SYCL devices as well as in host C++ code.
+///
+/// \ingroup sycl_api
+template <typename Type, int NumElements> class marray {
+  using DataT = Type;
+
+public:
+  using value_type = Type;
+  using reference = Type &;
+  using const_reference = const Type &;
+  using iterator = Type *;
+  using const_iterator = const Type *;
+
+private:
+  value_type m_Data[NumElements];
+
+  template <class...> struct conjunction : std::true_type {};
+  template <class B1, class... tail>
+  struct conjunction<B1, tail...>
+      : std::conditional<bool(B1::value), conjunction<tail...>, B1>::type {};
+
+  // TypeChecker is needed for (const argTN &... args) ctor to validate args.
+  template <typename T, typename DataT_>
+  struct TypeChecker : std::is_convertible<T, DataT_> {};
+
+  // Shortcuts for args validation in (const argTN &... args) ctor.
+  template <typename... argTN>
+  using EnableIfSuitableTypes = typename std::enable_if<
+      conjunction<TypeChecker<argTN, DataT>...>::value>::type;
+
+public:
+  marray() : m_Data{} {}
+
+  explicit marray(const Type &arg) {
+    for (int i = 0; i < NumElements; i++) {
+      m_Data[i] = arg;
+    }
```

Comment:
Just curious. Is there a reason not to use `std::fill`?

Reply:
I did not find usage of `std::fill` in similar cases in other places, thus avoided introducing new dependency, where it did not simplify much in code.